hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java

```diff
@@ -269,6 +269,14 @@ public SessionImpl(SessionFactoryImpl factory, SessionCreationOptions options) {
 		}
 	}
 
+	protected StatefulPersistenceContext createPersistenceContext() {
```

Comment:
Ah this is tricky; the new protected methods you've added are capturing "this", but since they are invoked from the constructor the reference to "this" isn't fully initialized yet.  This might seem to work but I find it a bit scary, especially as I expect you want to override the protected method from a subclass - this would lead to subclasss methods being run while the parent instance isn't initialized.

Could you explore some alternatives?

Reply:
@DavideD, FYI...

Reply:
I'm not a big fun of this solution either. In this case it works because all the constructor involved don't do anything complicated. I will check if I can find a different solution.

Reply:
I think I have an idea

Reply:
thanks :)   If there's no better way we can keep it of course.

Reply:
What about this? https://github.com/DavideD/hibernate-orm/commit/e4294f8e9b7bfefe5494e1cfc8606c69f2665a27

@gbadner Could you check that it works for the Rx PR, please? I've tried it on a different PR and it should be OK but better be sure. Thanks

Reply:
Ah, just noticed this discussion.

@DavideD the new approach in #3407 is just as bad. Well, actually, it's worse, really, since we're passing `this` to an arbitrary function, not just calling a method of the subclass.

If we *really* want to fix this properly the Good fix is to make an abstract superclass with `getPersistenceContext()` and `getActionQueue()` methods.

But I'm not sure that's worth it in this case.